client/lib/wp/sync-handler/index.js

```diff
@@ -0,0 +1,210 @@
+/**
+ * Module dependencies
+ */
+import config from 'config';
+import localforage from 'localforage';
+import Hashes from 'jshashes';
+import { isWhitelisted } from './whitelist-handler';
+
+import debugFactory from 'debug';
+const debug = debugFactory( 'calypso:sync-handler' );
+
+// defaults
+const defaults = config( 'sync-handler-defaults' );
+defaults.driver = localforage[ defaults.driver ];
+
+/**
+ * SyncHandler class
+ */
+export class SyncHandler {
+	/**
+	 * Create a SyncHandler instance
+	 *
+	 * @param {Object} [setup] - localforage sync setup
+	 * @param {Function} handler - wpcom handler
+	 *
+	 * @return {Function} sync-handler wrapper
+	 */
+	constructor( setup, handler ) {
+		if ( typeof setup === 'function' ) {
+			handler = setup;
+			setup = {};
+		}
+
+		this.setup = Object.assign( {}, defaults, setup );
+		this.reqHandler = handler;
+		return this.syncHandlerInstance( handler );
+	}
+
+	syncHandlerInstance( handler ) {
+		return ( params, callback ) => {
+			// detect and no-sync proxy connection request
+			if ( params.metaAPI && params.metaAPI.accessAllUsersBlogs ) {
+				debug( 'skip - non-sync -proxy-handler request detected' );
+				return this.reqHandler( params, callback );
+			}
+
+			const reqParams = Object.assign( {}, params );
+			const { path } = reqParams;
+
+			// whitelist barrier
+			if ( ! isWhitelisted( params ) ) {
+				debug( 'not whitelisted: skip %o request', path );
+				return handler( params, callback );
+			};
+
+			// response sent flag
+			let responseSent = false;
+
+			// generate an unique resource key
+			const key = this.generateKey( params );
+
+			debug( 'starting to get resource ...' );
+
+			/**
+			 * Send response immediately to the client
+			 * sendingt the given data
+			 *
+			 * @param {Object} localRecord - response stored locally
+			 */
+			const localResponse = localRecord => {
+				// let's be optimistic
+				if ( localRecord ) {
+					responseSent = true;
+
+					debug( '%o stored(%o). Let\'s be optimistic ...\n', path, localRecord );
+					//callback( null, localRecord.body );
+				} else {
+					debug( 'No data for %o\n', path );
+				}
+			};
+
+			/**
+			 * Handling Error getting data locally
+			 *
+			 * @param {Error} err - err trying to get the local record
+			 */
+			const recordErrorHandler = err => {
+				// @TODO improve error handling here
+				console.error( err );
+			};
+
+			/**
+			 * Fetch data from WP.com
+			 *
+			 * @return {Promise} promise
+			 */
+			const networkFetch = () => {
```

Comment:
Maybe pull these functions out of syncHandlerInstance for readability. Due to the need for a closure it might looks something like:
```
const createNetworkFetch = function( params ) {
    return function() {
       //...
    } 
}
```
And below it would be added like:
```
this.retrieveRecord( key )
      .then( createLocalResponse( callback ), recordErrorHandler )
      .then( createNetworkFetch( params ) )
      .then( createCacheResponse( callback ), createNetworkErrorHandler( callback ) );
```